src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/SourceLauncher.java

```diff
@@ -201,6 +201,7 @@ private Class<?> execute(MemoryContext context, String[] mainArgs)
         try {
             ClassLoader loader = context.newClassLoaderFor(parentLoader, firstClassName);
             firstClass = Class.forName(firstClassName, false, loader);
+            Thread.currentThread().setContextClassLoader(loader);
```

Comment:
Are you sure this it the right place? It might be, but seeing it in a try-catch is surprising. It should be set before the initializer for the main class runs, which might be here but checking.

Reply:
Would the constructor of `MemoryClassLoader` be a better place?

Because `context.newClassLoaderFor()` has two exit-points where an instance of `MemoryClassLoader` is created.

Reply:
The try-catch cares for `ClassNotFoundException` instances thrown by `Class.forName()` - which doesn't initialize the main class. But swapping the lines to read:
```java
            Thread.currentThread().setContextClassLoader(loader);
            firstClass = Class.forName(firstClassName, false, loader);
```
should better ensure that the TCCL is set before the initializer for the main class runs.

Reply:
I found a better way to move the `TCCL`-related above the try-catch block.

Reply:
Good, I was puzzled why newClassLoaderFor was declared to throw CNFE. Related question is do you know why it uses Class.forName without running the class initializer? I guess I expected it would load loader.loadClass(firstClassName) and let the initializer run (it's going to run a bit later anyway). Main thing is that the initializer may do anything and the TCCL needs to be set before that executes (which is what you have now).

Reply:
For reference, the single-file launcher `execute()` method had: `Class.forName(mainClassName, true, cl);`

IIRC, we decided to call with `false` here in order to delay initialization for when the first class is not the class with the main method - accounting for the new protocal that also tries to match the file name with a later type name within the file.